x/logic/keeper/interpreter.go

```diff
@@ -141,7 +118,86 @@
 	return i, userOutputBuffer, err
 }
 
+// whitelistBlacklistHookFn returns a hook function that checks if the given predicate is allowed to be executed.
+// The predicate is allowed if it is in the whitelist or not in the blacklist.
+func whitelistBlacklistHookFn(whitelist, blacklist []string) engine.HookFunc {
+	allowed := lo.Reduce(
+		lo.Filter(interpreter.RegistryNames,
+			util.Indexed(util.WhitelistBlacklistMatches(whitelist, blacklist, prolog2.PredicateMatches))),
+		func(agg *orderedmap.OrderedMap[string, struct{}], item string, _ int) *orderedmap.OrderedMap[string, struct{}] {
+			agg.Set(item, struct{}{})
+			return agg
+		},
+		orderedmap.New[string, struct{}](orderedmap.WithCapacity[string, struct{}](len(interpreter.RegistryNames))))
+
+	return func(opcode engine.Opcode, operand engine.Term, env *engine.Env) error {
+		if opcode != engine.OpCall {
+			return nil
+		}
+
+		predicateStringer, ok := operand.(fmt.Stringer)
+		if !ok {
+			return engine.SyntaxError(operand, env)
+		}
+
+		predicate := predicateStringer.String()
+
+		if interpreter.IsRegistered(predicate) {
+			if _, found := allowed.Get(predicate); !found {
+				return engine.PermissionError(
+					prolog2.AtomOperationExecute,
+					prolog2.AtomPermissionForbiddenPredicate,
+					engine.NewAtom(predicate),
+					env,
+				)
+			}
+		}
+		return nil
+	}
+}
+
+// gasMeterHookFn returns a hook function that consumes gas based on the cost of the executed predicate.
+func gasMeterHookFn(ctx context.Context, gasPolicy types.GasPolicy) engine.HookFunc {
+	sdkctx := sdk.UnwrapSDKContext(ctx)
+	gasMeter := meter.WithWeightedMeter(sdkctx.GasMeter(), nonNilNorZeroOrDefaultUint64(gasPolicy.WeightingFactor, defaultWeightFactor))
+
+	return func(opcode engine.Opcode, operand engine.Term, env *engine.Env) (err error) {
+		if opcode != engine.OpCall {
+			return nil
+		}
+
+		operandStringer, ok := operand.(fmt.Stringer)
+		if !ok {
+			return engine.SyntaxError(operand, env)
+		}
+
+		predicate := operandStringer.String()
+
+		cost := lookupCost(predicate, defaultPredicateCost, gasPolicy.PredicateCosts)
+
+		defer func() {
+			if r := recover(); r != nil {
+				switch rType := r.(type) {
+				case storetypes.ErrorOutOfGas:
+					err = errorsmod.Wrapf(
+						types.LimitExceeded, "out of gas: %s <%s> (%d/%d)",
+						types.ModuleName, rType.Descriptor, sdkctx.GasMeter().GasConsumed(), sdkctx.GasMeter().Limit())
+				default:
+					panic(r)
+				}
+			}
+		}()
+		gasMeter.ConsumeGas(cost, predicate)
+
+		return nil
+	}
+}
+
 func lookupCost(predicate string, defaultCost uint64, costs []types.PredicateCost) uint64 {
+	if !interpreter.IsRegistered(predicate) {
+		return defaultCost
+	}
+
```

Comment:
That’s exactly the intended behavior. Unregistered predicates are user-defined, and a default gas consumption needs to be applied to them.

However, it looks like we’re using a hard-coded constant when we should actually be referencing the module’s `default_predicate_cost` configuration value instead. It should be fixed.

Reply:
If you don't mind, I don’t want to address this issue in this PR, as I want to keep it iso-functionality for now. Could I ask you to open a ticket for this? We’ll handle it in a future update. Thanks!

Reply:
Awesome! Thanks.

Just one more thing—since you pointed out the performance issue with the lookupCost function, could you open a ticket to assess potential performance improvements for it? Thanks.

Reply:
You made my day.